### ISSUE TITLE
Code breaks if forward_batch_size is not a factor of batch_size

### ISSUE THREAD

**Issue by shubhlohiya** [lvwerra/trl#553]
On [this line](https://github.com/lvwerra/trl/blob/84c920903765deba2e3eb7f430db292b8f481001/trl/trainer/ppo_trainer.py#L885): 
```python
for i in range(int(bs / fbs)):
    ...
````
doesn't cover all mini batches if `bs` is not a multiple of `fbs`. This leads to the following ___example___ error in the `PPOTrainer.step` call:
```
ArrowInvalid: Column 2 named logprobs expected length 256 but got length 252
```

I would suggest changing this to `for i in range(math.ceil(bs / fbs))` or `for i in range(0, bs, fbs):` with appropriate changes to the slicing indices.

On the other hand, if this breaks the code elsewhere and it is necessary for `fbs` to be a factor of `bs`, please update docs and throw an error while defining the PPOConfig if this condition is violated.

**Comment by 1485840691-eng** [comment]
I personally tried to do the following changes on the 2 lines:

  1)   [this line](https://github.com/lvwerra/trl/blob/84c920903765deba2e3eb7f430db292b8f481001/trl/trainer/ppo_trainer.py#L885C5-L885C5)
        for i in range(int(bs / fbs)):  CHANGE TO    for i in range(math.ceil(bs / fbs)):
  2)   [this line](https://github.com/lvwerra/trl/blob/84c920903765deba2e3eb7f430db292b8f481001/trl/trainer/ppo_trainer.py#L902)
        for j in range(fbs):       CHNAGE TO     for j in range(len(query_batch)):
and use hello_world.py to try with 
     batch_size = 5
         mini_batch_size = 1
         mini_batch_size = 2    (when in this setting, the (input_ids, attention_mask) in input_kwargs, logits, query_batch will be in shape of [2:,], [2:, ], [1:, ])
         mini_batch_size = 3
         mini_batch_size = 5
no error popped up.

Following is the diff text

$ git diff trl/trainer/ppo_trainer.py
diff --git a/trl/trainer/ppo_trainer.py b/trl/trainer/ppo_trainer.py
index 8c44190..0cc3221 100644
--- a/trl/trainer/ppo_trainer.py
+++ b/trl/trainer/ppo_trainer.py
@@ -861,8 +861,8 @@ class PPOTrainer(BaseTrainer):
         all_logits = []
         all_masks = []
         all_values = []
-
-        for i in range(int(bs / fbs)):
+       import math
+       for i in range(math.ceil(bs / fbs)):
             input_kwargs = {key: value[i * fbs : (i + 1) * fbs] for key, value in model_inputs.items()}
             query_batch = queries[i * fbs : (i + 1) * fbs]
             response_batch = responses[i * fbs : (i + 1) * fbs]
@@ -879,7 +879,7 @@ class PPOTrainer(BaseTrainer):
             masks = torch.zeros_like(attention_mask)
             masks[:, :-1] = attention_mask[:, 1:]

-            for j in range(fbs):
+           for j in range(len(query_batch)):
                 if self.is_encoder_decoder:
                     # Decoder sentence starts always in the index 1 after padding in the Enc-Dec Models

**Comment by younesbelkada** [comment]
Hi everyone,
Thanks all for your suggestions, just made #563 that is based on this discussion
